File: vane-core/src/main/java/org/oddlama/vane/core/module/Module.java
package org.oddlama.vane.core.module;

import io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager;
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import org.bstats.bukkit.Metrics;
import org.bukkit.NamespacedKey;
import org.bukkit.OfflinePlayer;
import org.bukkit.attribute.Attribute;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.FishHook;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerFishEvent;
import org.bukkit.event.world.LootGenerateEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.loot.LootTables;
import org.bukkit.permissions.Permission;
import org.bukkit.permissions.PermissionAttachment;
import org.bukkit.permissions.PermissionDefault;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;
import org.oddlama.vane.annotation.VaneModule;
import org.oddlama.vane.annotation.config.ConfigBoolean;
import org.oddlama.vane.annotation.config.ConfigString;
import org.oddlama.vane.annotation.config.ConfigVersion;
import org.oddlama.vane.annotation.lang.LangVersion;
import org.oddlama.vane.annotation.persistent.Persistent;
import org.oddlama.vane.core.Core;
import org.oddlama.vane.core.LootTable;
import org.oddlama.vane.core.command.Command;
import org.oddlama.vane.core.config.ConfigManager;
import org.oddlama.vane.core.functional.Consumer1;
import org.oddlama.vane.core.lang.LangManager;
import org.oddlama.vane.core.persistent.PersistentStorageManager;
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static org.oddlama.vane.util.ResourceList.getResources;

public abstract class Module<T extends Module<T>> extends JavaPlugin implements Context<T>, org.bukkit.event.Listener {

	public final VaneModule annotation = getClass().getAnnotation(VaneModule.class);
	public Core core;
	public Logger log = getLogger();
	public ComponentLogger clog = getComponentLogger();
	private final String namespace = "vane_" + annotation.name().replaceAll("[^a-zA-Z0-9_]", "_");

	// Managers
	public ConfigManager configManager = new ConfigManager(this);
	public LangManager langManager = new LangManager(this);
	public PersistentStorageManager persistentStorageManager = new PersistentStorageManager(this);
	private boolean persistentStorageDirty = false;

	// Per module catch-all permissions
	public Permission permissionCommandCatchallModule;
	public Random random = new Random();

	// Permission attachment for console
	private List<String> pendingConsolePermissions = new ArrayList<>();
	public PermissionAttachment consoleAttachment;

	// Version fields for config, lang, storage
	@ConfigVersion
	public long configVersion;

	@LangVersion
	public long langVersion;

	@Persistent
	public long storageVersion;

	// Base configuration
	@ConfigString(def = "inherit", desc = "The language for this module. The corresponding language file must be named lang-{lang}.yml. Specifying 'inherit' will load the value set for vane-core.", metrics = true)
	public String configLang;

	@ConfigBoolean(def = true, desc = "Enable plugin metrics via bStats. You can opt-out here or via the global bStats configuration. All collected information is completely anonymous and publicly available.")
	public boolean configMetricsEnabled;

	// Context<T> interface proxy
	private ModuleGroup<T> contextGroup = new ModuleGroup<>(
			this,
			"",
			"The module will only add functionality if this is set to true.");

	@Override
	public void compile(ModuleComponent<T> component) {
		contextGroup.compile(component);
	}

	@Override
	public void addChild(Context<T> subcontext) {
		if (contextGroup == null) {
			// This happens, when contextGroup (above) is initialized and calls
			// compileSelf(),
			// while will try to register it to the parent context (us), but we fake that
			// anyway.
			return;
		}
		contextGroup.addChild(subcontext);
	}

	@Override
	public Context<T> getContext() {
		return this;
	}

	@SuppressWarnings("unchecked")
	@Override
	public T getModule() {
		return (T) this;
	}

	@Override
	public String yamlPath() {
		return "";
	}

	@Override
	public String variableYamlPath(String variable) {
		return variable;
	}

	@Override
	public boolean enabled() {
		return contextGroup.enabled();
	}

	// Callbacks for derived classes
	protected void onModuleLoad() {
	}

	public void onModuleEnable() {
	}

	public void onModuleDisable() {
	}

	public void onConfigChange() {
	}

	public void onGenerateResourcePack() throws IOException {
	}

	public final void forEachModuleComponent(final Consumer1<ModuleComponent<?>> f) {
		contextGroup.forEachModuleComponent(f);
	}

	// Loot modification
	private final Map<NamespacedKey, LootTable> additionalLootTables = new HashMap<>();

	// bStats
	public Metrics metrics;

	public Module() {
		// Get core plugin reference, important for inherited configuration
		// and shared state between vane modules
		if (this.getName().equals("vane-core")) {
			core = (Core) this;
		} else {
			core = (Core) getServer().getPluginManager().getPlugin("vane-core");
		}

		// Create per module command catch-all permission
		permissionCommandCatchallModule = new Permission(
				"vane." + getAnnotationName() + ".commands.*",
				"Allow access to all vane-" + getAnnotationName() + " commands",
				PermissionDefault.FALSE);
		registerPermission(permissionCommandCatchallModule);
	}

	/** The namespace used in resource packs */
	public final String namespace() {
		return namespace;
	}

	@Override
	public final void onLoad() {
		// Create data directory
		if (!getDataFolder().exists()) {
			getDataFolder().mkdirs();
		}

		onModuleLoad();
	}

	@Override
	public final void onEnable() {
		// Create console permission attachment
		consoleAttachment = getServer().getConsoleSender().addAttachment(this);
		for (var perm : pendingConsolePermissions) {
			consoleAttachment.setPermission(perm, true);
		}
		pendingConsolePermissions.clear();

		// Register in core
		core.registerModule(this);

		loadPersistentStorage();
		reloadConfiguration();

		// Schedule persistent storage saving every minute
		scheduleTaskTimer(
				() -> {
					if (persistentStorageDirty) {
						savePersistentStorage();
						persistentStorageDirty = false;
					}
				},
				60 * 20,
				60 * 20);
	}

	@Override
	public void onDisable() {
		disable();

		// Save persistent storage
		savePersistentStorage();

		// Unregister in core
		core.unregisterModule(this);
	}

	@Override
	public void enable() {
		if (configMetricsEnabled) {
			var id = annotation.bstats();
			if (id != -1) {
				metrics = new Metrics(this, id);
				configManager.registerMetrics(metrics);
			}
		}
		onModuleEnable();
		contextGroup.enable();
		registerListener(this);
	}

	@Override
	public void disable() {
		unregisterListener(this);
		contextGroup.disable();
		onModuleDisable();
		metrics = null;
	}

	@Override
	public void configChange() {
		onConfigChange();
		contextGroup.configChange();
	}

	@Override
	public void generateResourcePack(final ResourcePackGenerator pack) throws IOException {
		// Generate language
		final var pattern = Pattern.compile("lang-.*\\.yml");
		Arrays.stream(getDataFolder().listFiles((d, name) -> pattern.matcher(name).matches()))
				.sorted()
				.forEach(langFile -> {
					final var yaml = YamlConfiguration.loadConfiguration(langFile);
					try {
						langManager.generateResourcePack(pack, yaml, langFile);
					} catch (Exception e) {
						throw new RuntimeException(
								"Error while generating language for '" + langFile + "' of module " + getAnnotationName(),
								e);
					}
				});

		// Add files
		final var index = getResource("resource_pack/index");
		if (index != null) {
			try (final var reader = new BufferedReader(new InputStreamReader(index))) {
				String filePath;
				while ((filePath = reader.readLine()) != null) {
					final var content = getResource("resource_pack/" + filePath);
					pack.addFile(filePath, content);
				}
			} catch (IOException e) {
				log.log(Level.SEVERE, "Could not load resource pack index file of module " + getAnnotationName(), e);
			}
		}

		onGenerateResourcePack(pack);
		contextGroup.generateResourcePack(pack);
	}

	private boolean tryReloadConfiguration() {
		// Generate new file if not existing
		final var file = configManager.standardFile();
		if (!file.exists() && !configManager.generateFile(file, null)) {
			return false;
		}

		// Reload automatic variables
		return configManager.reload(file);
	}

	private void updateLangFile(String langFile) {
		final var file = new File(getDataFolder(), langFile);
		final var fileVersion = YamlConfiguration.loadConfiguration(file).getLong("Version", -1);
		long resourceVersion = -1;

		final var res = getResource(langFile);
		try (final var reader = new InputStreamReader(res)) {
			resourceVersion = YamlConfiguration.loadConfiguration(reader).getLong("Version", -1);
		} catch (IOException e) {
			log.log(Level.SEVERE, "Error while updating lang file '" + file + "' of module " + getAnnotationName(), e);
		}

		if (resourceVersion > fileVersion) {
			try {
				Files.copy(getResource(langFile), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
			} catch (IOException e) {
				log.log(Level.SEVERE, "Error while copying lang file '" + file + "' of module " + getAnnotationName(), e);
			}
		}
	}

	private boolean tryReloadLocalization() {
		// Copy all embedded lang files if their version is newer.
		getResources(getClass(), Pattern.compile("lang-.*\\.yml")).stream().forEach(this::updateLangFile);

		// Get configured language code
		var langCode = configLang;
		if ("inherit".equals(langCode)) {
			langCode = core.configLang;

			if (langCode == null) {
				// Core failed to load, so the server will be shutdown anyway.
				// Prevent an additional warning by falling back to en.
				langCode = "en";
			} else if ("inherit".equals(langCode)) {
				// Fallback to en in case 'inherit' is used in vane-core.
				langCode = "en";
			}
		}

		// Generate new file if not existing
		final var file = new File(getDataFolder(), "lang-" + langCode + ".yml");
		if (!file.exists()) {
			log.severe("Missing language file '" + file.getName() + "' for module " + getAnnotationName());
			return false;
		}

		// Reload automatic variables
		return langManager.reload(file);
	}

	public boolean reloadConfiguration() {
		boolean wasEnabled = enabled();

		if (!tryReloadConfiguration()) {
			// Force stop server, we encountered an invalid config file
			log.severe("Invalid plugin configuration. Shutting down.");
			getServer().shutdown();
			return false;
		}

		// Reload localization
		if (!tryReloadLocalization()) {
			// Force stop server, we encountered an invalid lang file
			log.severe("Invalid localization file. Shutting down.");
			getServer().shutdown();
			return false;
		}

		if (wasEnabled && !enabled()) {
			// Disable plugin if needed
			disable();
		} else if (!wasEnabled && enabled()) {
			// Enable plugin if needed
			enable();
		}

		configChange();
		return true;
	}

	public File getPersistentStorageFile() {
		// Generate new file if not existing
		return new File(getDataFolder(), "storage.json");
	}

	public void loadPersistentStorage() {
		// Load automatic persistent variables
		final var file = getPersistentStorageFile();
		if (!persistentStorageManager.load(file)) {
			// Force stop server, we encountered an invalid persistent storage file.
			// This prevents further corruption.
			log.severe("Invalid persistent storage. Shutting down to prevent further corruption.");
			getServer().shutdown();
		}
	}

	public void markPersistentStorageDirty() {
		persistentStorageDirty = true;
	}

	public void savePersistentStorage() {
		// Save automatic persistent variables
		final var file = getPersistentStorageFile();
		persistentStorageManager.save(file);
	}

	public void registerListener(Listener listener) {
		getServer().getPluginManager().registerEvents(listener, this);
	}

	public void unregisterListener(Listener listener) {
		HandlerList.unregisterAll(listener);
	}

	public String getAnnotationName() {
		return annotation.name();
	}

	public void registerCommand(Command<?> command) {
		LifecycleEventManager<Plugin> manager = this.getLifecycleManager();
		manager.registerEventHandler(LifecycleEvents.COMMANDS, event -> {
			event.registrar().register(command.getCommand(), command.langDescription.str(), command.getAliases());
		});
	}

	public void unregisterCommand(Command<?> command) {
		var bukkitCommand = command.getBukkitCommand();
		getServer().getCommandMap().getKnownCommands().values().remove(bukkitCommand);
		bukkitCommand.unregister(getServer().getCommandMap());
	}

	public void addConsolePermission(Permission permission) {
		addConsolePermission(permission.getName());
	}

	public void addConsolePermission(String permission) {
		if (consoleAttachment == null) {
			pendingConsolePermissions.add(permission);
		} else {
			consoleAttachment.setPermission(permission, true);
		}
	}

	public void registerPermission(Permission permission) {
		try {
			getServer().getPluginManager().addPermission(permission);
		} catch (IllegalArgumentException e) {
			log.log(Level.SEVERE, "Permission '" + permission.getName() + "' was already defined", e);
		}
	}

	public void unregisterPermission(Permission permission) {
		getServer().getPluginManager().removePermission(permission);
	}

	public LootTable lootTable(final LootTables table) {
		return lootTable(table.getKey());
	}

	public List<OfflinePlayer> getOfflinePlayersWithValidName() {
		return Arrays.stream(getServer().getOfflinePlayers())
				.filter(k -> k.getName() != null)
				.collect(Collectors.toList());
	}

	public LootTable lootTable(final NamespacedKey key) {
		var additionalLootTable = additionalLootTables.get(key);
		if (additionalLootTable == null) {
			additionalLootTable = new LootTable();
			additionalLootTables.put(key, additionalLootTable);
		}
		return additionalLootTable;
	}

	@EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
	public void onModuleLootGenerate(final LootGenerateEvent event) {
		final var lootTable = event.getLootTable();
		// Should never happen because according to the api this is @NotNull,
		// yet it happens for some people that copied their world from singleplayer to
		// the server.
		if (lootTable == null) {
			return;
		}
		final var additionalLootTable = additionalLootTables.get(lootTable.getKey());
		if (additionalLootTable == null) {
			return;
		}

		final var loc = event.getLootContext().getLocation();
		final var localRandom = new Random(
				random.nextInt() +
						(loc.getBlockX() & (0xffff << 16)) +
						(loc.getBlockY() & (0xffff << 32)) +
						(loc.getBlockZ() & (0xffff << 48)));
		additionalLootTable.generateLoot(event.getLoot(), localRandom);
	}

	@EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
	public void onModulePlayerCaughtFish(final PlayerFishEvent event) {
		// This is a dirty non-commutative way to apply fishing loot tables
		// that skews subtable probabilities,
		// consider somehow programmatically generating datapacks or
		// modifying loot tables directly instead.
		if (event.getState() != PlayerFishEvent.State.CAUGHT_FISH) {
			return;
		}
		if (event.getCaught() instanceof Item itemEntity) {
			final Player player = event.getPlayer();
			final FishHook hookEntity = event.getHook();
			final double playerLuck = player.getAttribute(Attribute.LUCK).getValue();
			final ItemStack rodStack = player.getInventory().getItem(event.getHand());
			final double rodLuck = rodStack.getEnchantmentLevel(Enchantment.LUCK_OF_THE_SEA);  // Can bukkit provide access to fishing_luck_bonus of 1.24 item component system?
			final double totalLuck = playerLuck + rodLuck;
			final double weightFish     =                               Math.max(0, 85 + totalLuck * -1);
			final double weightJunk     =                               Math.max(0, 10 + totalLuck * -2);
			final double weightTreasure = hookEntity.isInOpenWater() ? Math.max(0, 5 + totalLuck * 2) : 0;
			final double roll = random.nextDouble() * (weightFish + weightJunk + weightTreasure);
			NamespacedKey key;
			if (roll < weightFish) {
				key = LootTables.FISHING_FISH.getKey();
			} else if (roll < weightFish + weightJunk) {
				key = LootTables.FISHING_JUNK.getKey();
			} else {
				key = LootTables.FISHING_TREASURE.getKey();
			}
			final var additionalLootTable = additionalLootTables.get(key);
			if (additionalLootTable == null) {
				// Do not modify the caught item
				return;
			}
			final var newItem = additionalLootTable.generateOverride(new Random(random.nextInt()));
			if (newItem == null) {
				return;
			}
			itemEntity.setItemStack(newItem);
		}
	}
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/module/Module.kt
package org.oddlama.vane.core.module

import io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import net.kyori.adventure.text.logger.slf4j.ComponentLogger
import org.bstats.bukkit.Metrics
import org.bukkit.NamespacedKey
import org.bukkit.OfflinePlayer
import org.bukkit.attribute.Attribute
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Item
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerFishEvent
import org.bukkit.event.world.LootGenerateEvent
import org.bukkit.loot.LootTables
import org.bukkit.permissions.Permission
import org.bukkit.permissions.PermissionAttachment
import org.bukkit.permissions.PermissionDefault
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.java.JavaPlugin
import org.oddlama.vane.annotation.VaneModule
import org.oddlama.vane.annotation.config.ConfigBoolean
import org.oddlama.vane.annotation.config.ConfigString
import org.oddlama.vane.annotation.config.ConfigVersion
import org.oddlama.vane.annotation.lang.LangVersion
import org.oddlama.vane.annotation.persistent.Persistent
import org.oddlama.vane.core.Core
import org.oddlama.vane.core.LootTable
import org.oddlama.vane.core.command.Command
import org.oddlama.vane.core.config.ConfigManager
import org.oddlama.vane.core.functional.Consumer1
import org.oddlama.vane.core.lang.LangManager
import org.oddlama.vane.core.persistent.PersistentStorageManager
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator
import org.oddlama.vane.util.ResourceList
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger
import java.util.regex.Pattern
import kotlin.math.max

abstract class Module<T : Module<T?>?> : JavaPlugin(), Context<T?>, Listener {
    val annotation: VaneModule = javaClass.getAnnotation(VaneModule::class.java)!!
    var core: Core? = null
    var log: Logger = logger
    var clog: ComponentLogger = componentLogger
    private val namespace = "vane_" + annotation.name.replace("[^a-zA-Z0-9_]".toRegex(), "_")

    // Managers
    var configManager: ConfigManager = ConfigManager(this)
    var langManager: LangManager = LangManager(this)
    var persistentStorageManager: PersistentStorageManager = PersistentStorageManager(this)
    private var persistentStorageDirty = false

    // Per module catch-all permissions
    var permissionCommandCatchallModule: Permission?
    var random: Random = Random()

    // Permission attachment for console
    private val pendingConsolePermissions: MutableList<String> = mutableListOf()
    var consoleAttachment: PermissionAttachment? = null

    // Version fields for config, lang, storage
    @ConfigVersion
    var configVersion: Long = 0

    @LangVersion
    var langVersion: Long = 0

    @Persistent
    var storageVersion: Long = 0

    // Base configuration
    @ConfigString(
        def = "inherit",
        desc = "The language for this module. The corresponding language file must be named lang-{lang}.yml. Specifying 'inherit' will load the value set for vane-core.",
        metrics = true
    )
    var configLang: String? = null

    @ConfigBoolean(
        def = true,
        desc = "Enable plugin metrics via bStats. You can opt-out here or via the global bStats configuration. All collected information is completely anonymous and publicly available."
    )
    var configMetricsEnabled: Boolean = false

    // Context<T> interface proxy
    private val contextGroup: ModuleGroup<T?> = ModuleGroup<T?>(
        this,
        "",
        "The module will only add functionality if this is set to true.",
        false
    )

    override fun compile(component: ModuleComponent<T?>?) { contextGroup.compile(component) }

    override fun addChild(subcontext: Context<T?>?) {
        // Prevent adding the module's root group as a child of itself which would create
        // a self-referential cycle leading to infinite recursion on enable/disable.
        if (subcontext === contextGroup) return
        contextGroup.addChild(subcontext)
    }

    override val context: Context<T?>? get() = this

    @Suppress("UNCHECKED_CAST")
    override val module: T? get() = this as T

    override fun yamlPath(): String? = ""
    override fun variableYamlPath(variable: String?): String? = variable
    override fun enabled(): Boolean = contextGroup.enabled()

    // Callbacks for derived classes
    protected fun onModuleLoad() {}
    override fun onModuleEnable() {}
    override fun onModuleDisable() {}
    override fun onConfigChange() {}
    @Throws(IOException::class) fun onGenerateResourcePack() {}

    override fun forEachModuleComponent(f: Consumer1<ModuleComponent<*>?>?) {
        contextGroup.forEachModuleComponent(f)
    }

    // Loot modification
    private val additionalLootTables: MutableMap<NamespacedKey?, LootTable?> = mutableMapOf()

    // bStats
    var metrics: Metrics? = null

    init {
        // Get core plugin reference, important for inherited configuration
        // and shared state between vane modules
        core = if (this.name == "vane-core") this as Core
               else server.pluginManager.getPlugin("vane-core") as Core?

        // Create per module command catch-all permission
        permissionCommandCatchallModule = Permission(
            "vane.${annotationName}.commands.*",
            "Allow access to all vane-$annotationName commands",
            PermissionDefault.FALSE
        ).also { registerPermission(it) }

        // Compile the context group now that the Module initialization has finished
        contextGroup.compileSelf()
    }

    /** The namespace used in resource packs  */
    override fun namespace(): String = namespace

    override fun onLoad() {
        // Create data directory
        if (!dataFolder.exists()) dataFolder.mkdirs()
        onModuleLoad()
    }

    override fun onEnable() {
        // Create console permission attachment
        consoleAttachment = server.consoleSender.addAttachment(this).also { attachment ->
            pendingConsolePermissions.forEach { attachment.setPermission(it, true) }
            pendingConsolePermissions.clear()
        }

        // Register in core
        core!!.registerModule(this)
        loadPersistentStorage()
        reloadConfiguration()

        // Schedule persistent storage saving every minute
        scheduleTaskTimer(
            { if (persistentStorageDirty) { savePersistentStorage(); persistentStorageDirty = false } },
            (60 * 20).toLong(),
            (60 * 20).toLong()
        )
    }

    override fun onDisable() {
        disable()
        // Save persistent storage
        savePersistentStorage()
        // Unregister in core
        core!!.unregisterModule(this)
    }

    override fun enable() {
        if (configMetricsEnabled) {
            val id = annotation.bstats
            if (id != -1) {
                metrics = Metrics(this, id)
                configManager.registerMetrics(metrics)
            }
        }
        onModuleEnable()
        contextGroup.enable()
        registerListener(this)
    }

    override fun disable() {
        unregisterListener(this)
        contextGroup.disable()
        onModuleDisable()
        metrics = null
    }

    override fun configChange() {
        onConfigChange()
        contextGroup.configChange()
    }

    @Throws(IOException::class)
    override fun generateResourcePack(pack: ResourcePackGenerator?) {
        if (pack == null) return

        // Generate language files
        val langPattern = Pattern.compile("lang-.*\\.yml")
        val langFiles = dataFolder.listFiles { _, name -> name != null && langPattern.matcher(name).matches() }
            ?: emptyArray()
        langFiles.sorted().forEach { langFile ->
            val yaml = YamlConfiguration.loadConfiguration(langFile)
            try {
                langManager.generateResourcePack(pack, yaml, langFile)
            } catch (e: Exception) {
                throw RuntimeException(
                    "Error while generating language for '$langFile' of module $annotationName", e
                )
            }
        }

        // Add resource pack files listed in the embedded index
        getResource("resource_pack/index")?.let { index ->
            try {
                BufferedReader(InputStreamReader(index)).use { reader ->
                    reader.lineSequence().forEach { filePath ->
                        getResource("resource_pack/$filePath")?.let { content ->
                            pack.addFile(filePath, content)
                        } ?: log.log(Level.WARNING, "Missing resource 'resource_pack/$filePath' in module $annotationName")
                    }
                }
            } catch (e: IOException) {
                log.log(Level.SEVERE, "Could not load resource pack index file of module $annotationName", e)
            }
        }

        onGenerateResourcePack(pack)
        contextGroup.generateResourcePack(pack)
    }

    private fun tryReloadConfiguration(): Boolean {
        // Generate new file if not existing
        val file = configManager.standardFile()
        if (!file.exists() && !configManager.generateFile(file, null)) return false
        // Reload automatic variables
        return configManager.reload(file)
    }

    private fun updateLangFile(langFile: String) {
        val file = File(dataFolder, langFile)
        val fileVersion = YamlConfiguration.loadConfiguration(file).getLong("Version", -1)
        var resourceVersion = -1L

        getResource(langFile)?.let { res ->
            try {
                InputStreamReader(res).use { reader ->
                    resourceVersion = YamlConfiguration.loadConfiguration(reader).getLong("Version", -1)
                }
            } catch (e: IOException) {
                log.log(Level.SEVERE, "Error while updating lang file '$file' of module $annotationName", e)
            }
        }

        if (resourceVersion > fileVersion) {
            try {
                getResource(langFile)?.let { res ->
                    Files.copy(res, file.toPath(), StandardCopyOption.REPLACE_EXISTING)
                } ?: log.log(Level.WARNING, "Embedded lang resource '$langFile' missing for module $annotationName")
            } catch (e: IOException) {
                log.log(Level.SEVERE, "Error while copying lang file '$file' of module $annotationName", e)
            }
        }
    }

    private fun tryReloadLocalization(): Boolean {
        ResourceList.getResources(javaClass, Pattern.compile("lang-.*\\.yml"))
            .forEach { updateLangFile(it) }

        var langCode = configLang
        if (langCode == "inherit") {
            langCode = core!!.configLang ?: "en"
            if (langCode == "inherit") langCode = "en"
        }

        val file = File(dataFolder, "lang-$langCode.yml")
        if (!file.exists()) {
            log.severe("Missing language file '${file.name}' for module $annotationName")
            return false
        }
        return langManager.reload(file)
    }

    fun reloadConfiguration(): Boolean {
        val wasEnabled = enabled()

        if (!tryReloadConfiguration()) {
            // Force stop server, we encountered an invalid config file
            log.severe("Invalid plugin configuration. Shutting down.")
            server.shutdown()
            return false
        }

        // Reload localization
        if (!tryReloadLocalization()) {
            // Force stop server, we encountered an invalid lang file
            log.severe("Invalid localization file. Shutting down.")
            server.shutdown()
            return false
        }

        when {
            wasEnabled && !enabled() -> disable()
            !wasEnabled && enabled() -> enable()
        }

        configChange()
        return true
    }

    val persistentStorageFile: File
        get() = // Generate new file if not existing
            File(dataFolder, "storage.json")

    fun loadPersistentStorage() {
        // Load automatic persistent variables
        val file: File = this.persistentStorageFile
        if (!persistentStorageManager.load(file)) {
            // Force stop server, we encountered an invalid persistent storage file.
            // This prevents further corruption.
            log.severe("Invalid persistent storage. Shutting down to prevent further corruption.")
            server.shutdown()
        }
    }

    override fun markPersistentStorageDirty() { persistentStorageDirty = true }

    fun savePersistentStorage() {
        // Save automatic persistent variables
        val file: File = this.persistentStorageFile
        persistentStorageManager.save(file)
    }

    fun registerListener(listener: Listener) = server.pluginManager.registerEvents(listener, this)
    fun unregisterListener(listener: Listener) = HandlerList.unregisterAll(listener)

    val annotationName: String get() = annotation.name

    fun registerCommand(command: Command<*>) {
        val manager: LifecycleEventManager<Plugin> = lifecycleManager
        manager.registerEventHandler(LifecycleEvents.COMMANDS) { event ->
            event.registrar().register(command.command, command.langDescription.str(), command.getAliases())
        }
    }

    fun unregisterCommand(command: Command<*>) {
        val bukkitCommand = command.getBukkitCommand()
        server.commandMap.knownCommands.values.remove(bukkitCommand)
        bukkitCommand.unregister(server.commandMap)
    }

    fun addConsolePermission(permission: Permission) = addConsolePermission(permission.name)

    fun addConsolePermission(permission: String) {
        consoleAttachment?.setPermission(permission, true) ?: pendingConsolePermissions.add(permission)
    }

    fun registerPermission(permission: Permission) {
        try {
            server.pluginManager.addPermission(permission)
        } catch (e: IllegalArgumentException) {
            log.log(Level.SEVERE, "Permission '${permission.name}' was already defined", e)
        }
    }

    fun unregisterPermission(permission: Permission) = server.pluginManager.removePermission(permission)

    fun lootTable(table: LootTables): LootTable = lootTable(table.key)

    val offlinePlayersWithValidName: List<OfflinePlayer>
        get() = server.offlinePlayers.filter { it.name != null }

    fun lootTable(key: NamespacedKey?): LootTable =
        additionalLootTables.getOrPut(key) { LootTable() }!!

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun onModuleLootGenerate(event: LootGenerateEvent) {
        val additionalLootTable = additionalLootTables[event.lootTable.key] ?: return
        val loc = event.lootContext.location
        val localRandom = Random(
            (random.nextInt() +
                (loc.blockX and (0xffff shl 16)) +
                (loc.blockY and (0xffff shl 32)) +
                (loc.blockZ and (0xffff shl 48))).toLong()
        )
        additionalLootTable.generateLoot(event.loot, localRandom)
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun onModulePlayerCaughtFish(event: PlayerFishEvent) {
        if (event.state != PlayerFishEvent.State.CAUGHT_FISH) return
        val caughtItem = event.caught as? Item ?: return

        val player = event.player
        val playerLuck = player.getAttribute(Attribute.LUCK)!!.value
        val rodLuck = player.inventory.getItem(event.hand!!).getEnchantmentLevel(Enchantment.LUCK_OF_THE_SEA).toDouble()
        val totalLuck = playerLuck + rodLuck

        val weightFish     = max(0.0, 85 + totalLuck * -1)
        val weightJunk     = max(0.0, 10 + totalLuck * -2)
        val weightTreasure = if (event.hook.isInOpenWater) max(0.0, 5 + totalLuck * 2) else 0.0

        val roll = random.nextDouble() * (weightFish + weightJunk + weightTreasure)
        val key = when {
            roll < weightFish                  -> LootTables.FISHING_FISH.key
            roll < weightFish + weightJunk     -> LootTables.FISHING_JUNK.key
            else                               -> LootTables.FISHING_TREASURE.key
        }

        val additionalLootTable = additionalLootTables[key] ?: return
        val newItem = additionalLootTable.generateOverride(Random(random.nextInt().toLong())) ?: return
        caughtItem.itemStack = newItem
    }
}
